presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,211 @@
+            throw new ConfigurationException(ImmutableList.of(new Message("max-query-range-duration must be greater than query-chunk-size-duration")));
+        }
+
+        tableSupplier = Suppliers.memoizeWithExpiration(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)),
+                timeUnitsToSeconds(config.getCacheDuration()), TimeUnit.SECONDS);
+        MapType varcharMapType = (MapType) typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
+        this.config = config;
+        this.prometheusColumns = ImmutableList.of(
+                new PrometheusColumn("labels", varcharMapType),
+                new PrometheusColumn("timestamp", TimestampType.TIMESTAMP),
+                new PrometheusColumn("value", DoubleType.DOUBLE));
+    }
+
+    public Set<String> getSchemaNames()
+    {
+        return ImmutableSet.of("default");
+    }
+
+    private URI getPrometheusMetricsURI(PrometheusConnectorConfig config)
+            throws URISyntaxException
+    {
+        // endpoint to retrieve metric names from Prometheus
+        URI uri = config.getPrometheusURI();
+        return new URIBuilder()
+                .setScheme(uri.getScheme())
+                .setHost(uri.getAuthority())
+                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
+                .build();
+    }
+
+    public Set<String> getTableNames(String schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (schema.equals("default")) {
+            String status = (String) tableSupplier.get().get("status");
+            if (status.equals("success")) {
+                List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+                if (tableNames == null) {
+                    return ImmutableSet.of();
+                }
+                return tableNames.stream().collect(Collectors.toSet());
+            }
+            else {
+                //TODO this deals with success|error, there may be warnings to handle
+                return ImmutableSet.of();
+            }
+        }
+        return ImmutableSet.of();
+    }
+
+    public PrometheusTable getTable(String schema, String tableName)
+    {
+        requireNonNull(schema, "schema is null");
+        requireNonNull(tableName, "tableName is null");
+        if (schema.equals("default")) {
+            List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+            if (tableNames == null) {
+                return null;
+            }
+            if (!tableNames.contains(tableName)) {
+                return null;
+            }
+            PrometheusTable table = new PrometheusTable(
+                    tableName,
+                    prometheusColumns);
+            pushedDown.ifPresent(pushedDownPredicate -> table.setPredicate(Optional.of(pushedDownPredicate)));
+            return table;
+        }
+        else {
+            return null;
```

Comment:
return `Optional.empty()`

Reply:
Unless `TableNotFoundException` was also changed to implement `Supplier` I'm not sure that the change from `null` to `Optional.empty()` buys much because we can't make the usages particularly more functional.